crates/driver/src/domain/mempools.rs

```diff
@@ -104,11 +105,15 @@ impl Mempools {
         let result = async {
             loop {
                 // Wait for the next block to be mined or we time out.
-                if tokio::time::timeout_at(deadline, block_stream.next())
-                    .await
-                    .is_err()
-                {
-                    tracing::info!(?hash, "tx not confirmed in time, cancelling");
+                let current_block = self.ethereum.current_block().borrow().number;
+                if current_block > submission_deadline {
```

Comment:
It looks like it doesn't make sense to check this condition more often than a new block is mined. Since there is access to `block_stream` already, it can be used instead of loop like
```
while let Some(block) = block_stream.next().await {
...
}
```